codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -167,11 +167,29 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                 }
             }
 
-            // Pass headers to the flow-controller so it can maintain their sequence relative to DATA frames.
-            flowController().addFlowControlled(stream,
-                    new FlowControlledHeaders(stream, headers, streamDependency, weight, exclusive, padding,
-                            endOfStream, promise));
-            return promise;
+            // Trailing headers must go through flow control if there are other frames queued in flow control
+            // for this stream.
+            Http2RemoteFlowController flowController = flowController();
+            if (!endOfStream || !flowController.hasFlowControlled(stream)) {
```

Comment:
This is an option, however I would actually favor this approach as opposed to using properties on the stream. I would rather track less state, and have a smaller API on the stream if possible ... the flow controller is already queuing state and we can infer if these are headers or trailers because of protocol messaging restrictions.